qiskit_experiments/curve_analysis/curve_analysis.py

```diff
@@ -150,6 +217,248 @@ def _data(
+        curve_data: CurveData,
+    ) -> CurveData:
+        """Postprocessing for the processed dataset.
+
+        Args:
+            curve_data: Processed dataset created from experiment results.
+
+        Returns:
+            Formatted data.
+        """
+        # take average over the same x value by keeping sigma
+        data_allocation, xdata, ydata, sigma, shots = multi_mean_xy_data(
+            series=curve_data.data_allocation,
+            xdata=curve_data.x,
+            ydata=curve_data.y,
+            sigma=curve_data.y_err,
+            shots=curve_data.shots,
+            method="shots_weighted",
+        )
+
+        # sort by x value in ascending order
+        data_allocation, xdata, ydata, sigma, shots = data_sort(
+            series=data_allocation,
+            xdata=xdata,
+            ydata=ydata,
+            sigma=sigma,
+            shots=shots,
+        )
+
+        return CurveData(
+            x=xdata,
+            y=ydata,
+            y_err=sigma,
+            shots=shots,
+            data_allocation=data_allocation,
+            labels=curve_data.labels,
+        )
+
+    def _generate_fit_guesses(
+        self,
+        user_opt: FitOptions,
+        curve_data: CurveData,  # pylint: disable=unused-argument
+    ) -> Union[FitOptions, List[FitOptions]]:
+        """Create algorithmic guess with analysis options and curve data.
+
+        Args:
+            user_opt: Fit options filled with user provided guess and bounds.
+            curve_data: Formatted data collection to fit.
+
+        Returns:
+            List of fit options that are passed to the fitter function.
+        """
+        return user_opt
+
+    def _run_curve_fit(
+        self,
+        curve_data: CurveData,
+        models: List[lmfit.Model],
+    ) -> CurveFitResult:
+        """Perform curve fitting on given data collection and fit models.
+
+        Args:
+            curve_data: Formatted data to fit.
+            models: A list of LMFIT models that are used to build a cost function
+                for the LMFIT minimizer.
+
+        Returns:
+            The best fitting outcome with minimum reduced chi-squared value.
+        """
+        unite_parameter_names = []
+        for model in models:
+            # Seems like this is not efficient looping, but using set operation sometimes
+            # yields bad fit. Not sure if this is an edge case, but
+            # `TestRamseyXY` unittest failed due to the significant chisq value
+            # in which the least_square fitter terminates with `xtol` rather than `ftol`
+            # condition, i.e. `ftol` condition indicates termination by cost function.
+            # This code respects the ordering of parameters so that it matches with
+            # the signature of fit function and it is backward compatible.
+            # In principle this should not matter since LMFIT maps them with names
+            # rather than index. Need more careful investigation.
```

Comment:
Do you want to open issue for this?

Reply:
I think this looping is not significant overhead compared with other lines (I've done some profiling). So this is okey for now with this comment.